lib/msal-common/src/client/DeviceCodeClient.ts

```diff
@@ -0,0 +1,190 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { DeviceCodeResponse } from "../response/DeviceCodeResponse";
+import { BaseClient } from "./BaseClient";
+import { DeviceCodeRequest } from "../request/DeviceCodeRequest";
+import { Authority } from "../authority/Authority";
+import { ClientAuthError } from "../error/ClientAuthError";
+import { RequestUtils } from "../utils/RequestUtils";
+import { RequestValidator } from "../request/RequestValidator";
+import { Constants, GrantType } from "../utils/Constants";
+import { Configuration } from "../config/Configuration";
+import { TimeUtils } from "../utils/TimeUtils";
+
+/**
+ * OAuth2.0 Device code client
+ */
+export class DeviceCodeClient extends BaseClient {
+
+    private authority: Authority;
+
+    constructor(configuration: Configuration){
+        super({
+            authOptions: configuration.authOptions,
+            systemOptions: configuration.systemOptions,
+            loggerOptions: configuration.loggerOptions,
+            storageInterface: configuration.storageInterface,
+            networkInterface: configuration.networkInterface,
+            cryptoInterface: configuration.cryptoInterface
+        });
+    }
+
+    /**
+     * Gets device code from device code endpoint, calls back to with device code response, and
+     * polls token endpoint to exchange device code for tokens
+     * @param request
+     */
+    public async acquireToken(request: DeviceCodeRequest): Promise<string> {
+        this.authority = await this.createAuthority(request.authority);
+        const deviceCodeResponse: DeviceCodeResponse = await this.getDeviceCode(request);
+        request.deviceCodeCallback(deviceCodeResponse);
+        return this.acquireTokenWithDeviceCode(request, deviceCodeResponse);
+    }
+
+    /**
+     * Creates device code request and executes http GET
+     * @param request
+     */
+    private async getDeviceCode(request: DeviceCodeRequest): Promise<DeviceCodeResponse> {
+
+        const deviceCodeUrl = this.createDeviceCodeUrl(request);
+        const headers: Map<string, string> = new Map<string, string>();
+        RequestUtils.addLibrarydataHeaders(headers);
+
+        return this.executeGetRequest(deviceCodeUrl, headers);
+    }
+
+    /**
+     * Executes GET request to device code endpoint
+     * @param deviceCodeUrl
+     * @param headers
+     */
+    private async executeGetRequest(deviceCodeUrl: string, headers: Map<string, string>): Promise<DeviceCodeResponse>{
+
+        let deviceCodeResponse;
```

Comment:
return this.networkClient.sendGetRequestAsync(...)
    .catch(err => {
          //log error and then
          throw err;    
})